lightning/src/ln/channelmanager.rs

```diff
@@ -764,6 +767,16 @@ where
 
 	/// See `ChannelManager` struct-level documentation for lock order requirements.
 	pending_events: Mutex<Vec<events::Event>>,
+	/// When a peer disconnects but still has channels, the peer's `peer_state` entry in the
+	/// `per_peer_state` is not removed by the `peer_disconnected` function. If the channels of
+	/// to that peer is later closed while still being disconnected (i.e. force closed), we
+	/// therefore need to remove the peer from `peer_state` separately.
+	/// To avoid having to take the `per_peer_state` `write` lock once the channels are closed, we
+	/// instead store such peers awaiting removal in this field, and remove them on a timer to
+	/// limit the negative effects on parallelism as much as possible.
+	///
+	/// See `ChannelManager` struct-level documentation for lock order requirements.
+	pending_peers_awaiting_removal: Mutex<HashSet<PublicKey>>,
```

Comment:
Let me know if we're better of just taking the write lock of the `per_peer_state` after force closing a channel and we're already disconnected with no channels left :).